### PR TITLE
Standalone PR creates task link to relevant ticket

### LINKED ISSUE #78: Standalone PR links parent Epic
When standalone PR is created and links a single Epic, set the Epic as a parent

Depends on #77

### LINKED ISSUE #79: Standalone PR creates subtask
When standalone PR is created and links a single issue that is not epic and is not a subtask, create a subtask with the mentioned parent

Depends on https://github.com/SonarSource/gh-action-lt-backlog/issues/77

### LINKED ISSUE #80: Standalone PR links first epic
When standalone PR is created and links multiple issues, iterate issues and:
* Use 1st epic and use it as a parent
* Create issue without a parent otherwise

Depends on #77 and #78

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #78
Fixes #79

### CODE REVIEW

File: lib/OctokitAction.ts [hunk 1]
```diff
@@ -34,8 +34,8 @@ export abstract class OctokitAction extends Action {
     return this.graphqlWithAuth(query);
   }
 
-  protected getInput(name: string): string {
-    return core.getInput(name);
+  protected getInput(name: string, required: boolean = false): string {
+    return core.getInput(name, { required });
```
Comment (pavel-mikula-sonarsource) [inline lib/OctokitAction.ts:38]: Why this change? Required parameters are defined in the `action.yml`, and if anything is optional, it should have a default value there. And then, inside TS, we don't need to care.

File: PullRequestCreated/PullRequestCreated.ts [hunk 2]
```diff
@@ -16,6 +22,25 @@ class PullRequestCreated extends OctokitAction {
   private shouldCreateIssue(pr: PullRequest): boolean {
     return pr != null && !JIRA_ISSUE_PATTERN.test(pr.title);
   }
+
+  private async newIssueTypeAndParent(pr: PullRequest): Promise<Issue> {
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:26]: This, or something that is aligned with the new class name
```suggestion
  private async newIssueParameters(pr: PullRequest): Promise<...fixme...> {
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
 
 class PullRequestCreated extends OctokitAction {
   protected async execute(): Promise<void> {
     const pr = await this.getPullRequest(this.payload.pull_request.number);
     if (this.shouldCreateIssue(pr)) {
-      const projectKey = this.getInput('jira-project');
-      const issueKey = await this.jira.createIssue(projectKey, JIRA_TASK_ISSUE, pr.title);
+      const issue = await this.newIssueTypeAndParent(pr);
+      const projectKey = this.getInput('jira-project', true);
+      const issueKey = await this.jira.createIssue(projectKey, issue.type, pr.title, { parent: { key: issue.parent } });
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:16]: `issueKey` can be null if something goes wrong. We should add `if(issueKey != null)` before updating the title

File: PullRequestCreated/PullRequestCreated.ts [hunk 4]
```diff
@@ -16,6 +22,25 @@ class PullRequestCreated extends OctokitAction {
   private shouldCreateIssue(pr: PullRequest): boolean {
     return pr != null && !JIRA_ISSUE_PATTERN.test(pr.title);
   }
+
+  private async newIssueTypeAndParent(pr: PullRequest): Promise<Issue> {
+    const mentionnedIssues = this.findMentionnedIssues(pr);
+    console.log(`Found ${mentionnedIssues.length} mentionned issues: ${mentionnedIssues}`);
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:28]: The count is redundant
```suggestion
    console.log(`Found mentioned issues: ${mentionnedIssues}`);
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 5]
```diff
@@ -16,6 +22,25 @@ class PullRequestCreated extends OctokitAction {
...
+    let parent = mentionnedIssues.length === 1 ? await this.jira.getIssue(mentionnedIssues[0]) : null;
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:29]:
```suggestion
    const parent = mentionnedIssues.length === 1 ? await this.jira.getIssue(mentionnedIssues[0]) : null;
```

File: PullRequestCreated/PullRequestCreated.ts [hunk 6]
```diff
@@ -16,6 +22,25 @@ class PullRequestCreated extends OctokitAction {
   private shouldCreateIssue(pr: PullRequest): boolean {
     return pr != null && !JIRA_ISSUE_PATTERN.test(pr.title);
   }
+
+  private async newIssueTypeAndParent(pr: PullRequest): Promise<Issue> {
+    const mentionnedIssues = this.findMentionnedIssues(pr);
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:27]: Burguer 😆 
```suggestion
    const mentionedIssues = this.findMentionedIssues(pr);
```

File: PullRequestCreated/PullRequestCreated.ts (hunk 5 above)
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:29]: This logic does not respect "first Epic if there are multiple issues mentioned"

File: PullRequestCreated/PullRequestCreated.ts [hunk 7]
```diff
@@ -16,6 +22,25 @@ class PullRequestCreated extends OctokitAction {
...
+    console.log(`Found ${mentionnedIssues.length} mentionned issues: ${mentionnedIssues}`);
+    let parent = mentionnedIssues.length === 1 ? await this.jira.getIssue(mentionnedIssues[0]) : null;
+    console.log(`Parent issue: ${parent?.key} (${parent?.fields.issuetype.name})`);
+    switch (parent?.fields.issuetype.name) {
+      case 'Epic':
+        return { type: 'Task', parent: parent.key };
+      case 'Task':
+        return { type: 'Sub-task', parent: parent.key };
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:35]: This is wrong, it's not `Task`, it's any issue except subtasks and epics. Imagine `NewFeature`, `Improvement`, `False-Positive`, `False-Negative`...

File: PullRequestCreated/PullRequestCreated.ts (hunk 5 above)
Comment (sebastien-marichal) [reply PullRequestCreated/PullRequestCreated.ts:29]: This is a separate PR: #106

File: PullRequestCreated/ReadMe.md [hunk 8]
```diff
@@ -3,7 +3,12 @@
 Upon pull request creation, create a Jira ticket if no tickets are mentioned in the title.
 
 Depending on the content of the PR description, this action will create different types of Jira tickets:
-- A `Task` with no parent if there is no ticket mentioned in the description
+- A `Task` with no parent
+  - if there is no ticket mentioned in the description
+  - or, there is no `Epic` mentioned when multiple tickets are mentioned in the description
+- A `Task` with the mentioned `Epic` as a parent
+- A `Task` with the first `Epic` mentioned as a parent if there are multiple tickets mentioned
+- A `Sub-task` with the mentioned `Task` as a parent
```
[the remark this answers is not in the document]
Comment (sebastien-marichal) [reply PullRequestCreated/ReadMe.md:11]: IMO, That would make sense to take the first non-Sub-task ticket.

File: lib/OctokitAction.ts (hunk 1 above)
Comment (sebastien-marichal) [reply lib/OctokitAction.ts:38]: How does it work?
I made the change because I did not update the workflow after we moved from having only `JIRA_TOKEN` to having both `JIRA_USER` and `JIRA_TOKEN`.
If I don't set those inputs, which are required, I am having an error during the action execution:
```
401 (Unauthorized): You do not have permission to create issues in this project.
```

I am expecting to have an issue before and a failed workflow then.
It seems that even with the required field in TS, it does not work.

Comment (sebastien-marichal) [reply lib/OctokitAction.ts]: I reverted this change

Comment (pavel-mikula-sonarsource) [reply lib/OctokitAction.ts]: If you don't send parameter that action.yml requires, it should fail quickly. But I didn't test it.
If you send useless parameters that are not expected by action.yml, you'll get a warning.

File: PullRequestCreated/PullRequestCreated.ts [hunk 9]
```diff
@@ -1,21 +1,52 @@
...
       const projectKey = this.getInput('jira-project');
-      const issueKey = await this.jira.createIssue(projectKey, JIRA_TASK_ISSUE, pr.title);
-      await this.updatePullRequestTitle(this.payload.pull_request.number, `${issueKey} ${pr.title}`);
-      await this.updatePullRequestDescription(this.payload.pull_request.number, `${issueKey}\n\n${pr.body || ''}`);
+      const issueKey = await this.jira.createIssue(projectKey, issueParameter.type, pr.title, { parent: { key: issueParameter.parent } });
+      if (issueKey != null) {
+        await this.updatePullRequestTitle(this.payload.pull_request.number, `${issueKey} ${pr.title}`);
+        await this.updatePullRequestDescription(this.payload.pull_request.number, `${issueKey}\n\n${pr.body || ''}`);
+      } else {
+        console.log('Unable to update PR title and description');
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/PullRequestCreated.ts:21]: To follow the logging design, let's do this login inside `createIssue`. Like that, it will be done once for all and we won't need to bother every caller with his own logging.
```suggestion
```

File: PullRequestCreated/ReadMe.md [hunk 10]
```diff
@@ -3,7 +3,9 @@
 Upon pull request creation, create a Jira ticket if no tickets are mentioned in the title.
 
 Depending on the content of the PR description, this action will create different types of Jira tickets:
-- A `Task` with no parent if there is no ticket mentioned in the description
+- A `Task` with the mentioned `Epic` as a parent
+- A `Sub-task` with the first ticket mentioned as a parent, except `Sub-task`
```
Comment (pavel-mikula-sonarsource) [inline PullRequestCreated/ReadMe.md:7]: The docs is currently out of date, as the "first" will be done in #106. It can probably stay that way for the moment

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
